### PR TITLE
[PowerPC] improve P10 store forwarding on P7 scalar to vector

### PR DESCRIPTION
**Pull request by RolandF77**
Commercial applications are often built for the oldest supported processor, even though they will usually be run on newer processors. Try to make code built for P7 that has scalar to vector operations run better on P10 by making a small change to the P7 scalar to vector code to allow hardware store forwarding on P10.

### CODE REVIEW

File: llvm/lib/Target/PowerPC/PPCISelLowering.cpp [hunk 1]
```diff
@@ -11479,9 +11492,33 @@ SDValue PPCTargetLowering::LowerSCALAR_TO_VECTOR(SDValue Op,
   EVT PtrVT = getPointerTy(DAG.getDataLayout());
   SDValue FIdx = DAG.getFrameIndex(FrameIdx, PtrVT);
 
+  SDValue Val = Op.getOperand(0);
+  EVT ValVT = Val.getValueType();
+  // P10 hardware store forwarding requires that a single store contains all
+  // the data for the load. P10 is able to merge a pair of adjacent stores. Try
```
Comment (chenzheng1030) [inline llvm/lib/Target/PowerPC/PPCISelLowering.cpp:11498]: For `adjacent stores` in P10 hardware store forwarding, do you mean there should be no instructions between these two stores in the final assembly? I tend to "no" as we have OOO on P10. But if the answer is yes, maybe we need a pseudo instruction to pack these two stores together and expand them after the schedulers... The chain in the codes can only guarantee the first store is before the second store.

Comment (RolandF77) [reply llvm/lib/Target/PowerPC/PPCISelLowering.cpp:11498]: Adjacent here means that the storage targets modified by the two stores have to be adjacent in memory. Other instructions can be between them, but not other stores.

Comment (chenzheng1030) [reply llvm/lib/Target/PowerPC/PPCISelLowering.cpp:11498]: Thanks for explanation. Got it.

**Review by chenzheng1030 — APPROVED** [review]
LGTM. Thanks for this "interesting" exploitation.

**Comment by RolandF77** [comment]
Passed build Linux x64. 1 Clang FE fail Win x64. I doubt PPC BE is to blame.

**Pull request merged**